File: Go/CountingSort.go
package main

import (
	"fmt"
	"math"
)

func main() {
	// Input int array
	array := []int{3, 82713871, 42, -123142, 35, 2, 12, -3, -4, -1, 3}
	sortedArray := CountingSort(array)
	fmt.Println(sortedArray)
}

/* 
Implements Counting Sort on []int and returns sorted []int
Input: []int 	eg: [3 82713871 42 -123142 35 2 12 -3 -4 -1 3]
Output: []int 	eg: [-123142 -4 -3 -1 2 3 3 12 35 42 82713871]
Time Complexity: O(n + k) where n is the number of elements in array and k is the range of input
Space Complexity: O(n + k) Auxilary array
*/
func CountingSort(arr []int) []int {
	/*
	Usual approach for implementing Counting Sort would be to directly use index
	from 0 to max element in the array but it fails while dealing with negative numbers.
	Therefore, what we do here is that we take min and max of the given array and map them
	to the array of index starting from 0 to min + max
	*/

	min := math.MaxInt64
	max := math.MinInt64
	
	// Find min and max
	for _, v := range arr {
		if min > v {
			min = v
		}
		if max < v {
			max = v
		}
	}

	// Initialze array to store counts
	count := make([]int, max - min + 1)

	for _, x := range arr {
		count[x-min]++
	}

	sortedArray := make([]int, len(arr))

	i := 0
	for k, v := range count {
		// Insert all v occurences of element k in empty array
		for v > 0 {
			sortedArray[i] = k + min
			v--
			i++
		}
	}

	return sortedArray
}

File: Python/CountingSort.py
#Python Program to Sort a List using Counting Sort Algorithm

def counting_sort(ar):
    m = int(max(ar)) + 1
    count = [0] * m   #Count List, it is initialised to a list of size one more than the largest list item

    #This for loop computes the frequency of each list item and stores it in the count list
    for a in ar:
        count[a] += 1

    i = 0


    for a in range(m):
        for c in range(count[a]):
            ar[i] = a
            i += 1

    return ar

l = []
n = int(input("Enter number of elements: "))

for i in range(0, n):
    e = int(input())

    l.append(e)

print("Initial List:",l)
print("Sorted List:",counting_sort(l))


'''
Sample input:
Enter number of elements: 5
1
5
8
1
4

Sample Output:
Initial List: [1,5,8,1,4]
Sorted List: [1,1,4,5,8]

Time Complexity is:O(n+max)
Space Complexity is:O(n+max)
                    where max: largest number in the input list
                    and n: size of input list

'''
